mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java

```diff
@@ -0,0 +1,93 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.cassandra.mail;
+
+import static com.datastax.driver.core.querybuilder.QueryBuilder.delete;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.eq;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.in;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.insertInto;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.select;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.apache.james.backends.cassandra.utils.CassandraUtils;
+import org.apache.james.mailbox.cassandra.CassandraId;
+import org.apache.james.mailbox.cassandra.table.CassandraAnnotationTable;
+import org.apache.james.mailbox.model.MailboxAnnotation;
+
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+
+public class CassandraAnnotationMapper {
+
+    private final CassandraId mailboxId;
+    private final Session session;
+
+    public CassandraAnnotationMapper(CassandraId mailboxId, Session session) {
+        this.mailboxId = mailboxId;
+        this.session = session;
+    }
+
+    public List<MailboxAnnotation> getAnnotations() {
+        return getResult(getStoredAnnotationRow());
+    }
+
+    public List<MailboxAnnotation> getAnnotations(List<String> keys) {
+        return getResult(getStoredAnnotationRow(keys));
+    }
+
+    public void deleteAnnotation(String key) {
+        session.execute(delete().from(CassandraAnnotationTable.TABLE_NAME)
+            .where(eq(CassandraAnnotationTable.MAILBOX_ID, mailboxId.asUuid()))
+            .and(eq(CassandraAnnotationTable.KEY, key)));
+    }
+
+    public void insertAnnotation(MailboxAnnotation mailboxAnnotation) {
```

Comment:
```
public static MailboxAnnotation nil(String key) {
        return new MailboxAnnotation(key, Optional.<String> absent());
    }
```
Nil is an instance of MailboxAnnotation. So it is legal to pass it to insertAnnotation. What other layers do is another subject, but we can't rely on layer coupling and say "we know the caller has a if to delete or insert based on the value in MailboxAnnotation".
If your method takes a MailboxAnnotation, be prepared to get all legal values of this type. Nil is a legal value, but when you will dereference the optional part, it will break.
You have to solutions then : 
1. you make a runtime preconditions at method start to fail as soon as possible
2. you make sure the input type restricts input values to some valid ones
The second one is what I call a typing strategy. For exemple, you could have : 
- `interface MailboxAnnotation`
- `class Nil implements MailboxAnnotation` (and Nil would be a singleton)
- `class KeyValueMailboxAnnotation implement MailboxAnnotation`
Then you could change `insertAnnotation(MailboxAnnotation mailboxAnnotation)` into `insertAnnotation(KeyValueMailboxAnnotation mailboxAnnotation)` and you'r done.
Is it clear enough ?